app/server/lib/ActiveDoc.ts

```diff
@@ -1012,6 +1023,49 @@ export class ActiveDoc extends EventEmitter {
     throw new ApiError(`Unsupported archive format ${format}`, 400);
   }
 
+  @ActiveDoc.keepDocOpen
+  public async addMissingFilesFromArchive(docSession: OptDocSession,
+                                          tarFile: stream.Readable): Promise<ArchiveUploadResult> {
+    // TODO - Check if this is the right access level for this.
+    if(!await this.isOwner(docSession)) {
+      throw new ApiError("Insufficient access to upload an attachment archive", 403);
+    }
+
+    const fallbackStoreId = this._getDocumentSettings().attachmentStoreId;
+    const results = {
+      added: 0,
+      errored: 0,
+      unused: 0,
+    };
+
+    await unpackTarArchive(tarFile, async (file) => {
+      try {
+        const fileName = path.basename(file.path);
+        const fileHash = file.path.split("_")[0];
+        const fileExt = path.extname(fileName);
+        const fileIdent = `${fileHash}${fileExt}`;
+        const isAdded = await this._attachmentFileManager.addMissingFileData(
+          fileIdent,
+          file.data,
+          fallbackStoreId,
+        );
+        if (isAdded) {
+          results.added += 1;
+        } else {
+          results.unused += 1;
+        }
+      } catch (err) {
+        results.errored += 1;
+        if (err instanceof MismatchedFileHashError) {
+          this._log.warn(docSession, `Failed to upload attachment: ${err.message}`);
+        }
+        this._log.error(docSession, `Failed to upload attachment: ${err}`);
```

Comment:
Should this also just be a warning?

Reply:
There's no particular standard. This seems ok. The log isn't very structured for ingestion, but that's ok.

Reply:
I wonder about maybe returning the stringified errors in a list in the results, since otherwise the problem(s) will be very opaque to user. Optional, easy to add later.

Reply:
I was hesitant about doing that, just in case they're cryptic and unhelpful runtime errors (like the classic "access property of undefined"). Easy either way from my perspective.